File: Pattern-DynamicProgramming/LongestCommonSubstring/minimumDeletionsToMakeSequenceSorted/bottomUp.js
const findMinimumDeletions = function (nums) {
  return nums.length - findLISLength(nums);
};

const findLISLength = function (nums) {
  const dp = Array(nums.length).fill(0);

  dp[0] = 1;
  let max = 1;

  for (let i = 1; i < nums.length; i++) {
    dp[i] = 1;

    for (let j = 0; j < i; j++) {
      if (nums[i] > nums[j] && dp[i] <= dp[j]) {
        dp[i] = dp[j] + 1;
        max = Math.max(max, dp[i]);
      }
    }
  }
  return max;
};

// Time complexity is O(n^2)

// The space complexity is O(n)

console.log(
  `Minimum deletion needed: ---> ${findMinimumDeletions([
    4, 2, 3, 6, 10, 1, 12,
  ])}`
);
console.log(
  `Minimum deletion needed: ---> ${findMinimumDeletions([-4, 10, 3, 7, 15])}`
);
console.log(
  `Minimum deletion needed: ---> ${findMinimumDeletions([3, 2, 1, 0])}`
);

File: Pattern-DynamicProgramming/LongestCommonSubstring/longestRepeatingSubsequence/bottomUp.py
def find_LRS_length(str):
    n = len(str)

    dp = [[0 for _ in range(n+1)] for _ in range(n+1)]
    maxLen = 0

    for i in range(1,n+1):
        for j in range(1,n+1):
            if str[i-1] == str[j-1] and i != j:
                dp[i][j] = 1 + dp[i-1][j-1]
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])
            maxLen = max(maxLen,dp[i][j])
    return maxLen

# Time complexity is O(n^2)

# Space complexity is O(n^2)

def main():
  print(find_LRS_length("tomorrow"))
  print(find_LRS_length("aabdbcec"))
  print(find_LRS_length("fmff"))


main()